api/src/main/java/org/openmrs/api/impl/ConceptServiceImpl.java

```diff
@@ -1981,6 +1985,77 @@ public ConceptAttribute getConceptAttributeByUuid(String uuid) {
 	public boolean hasAnyConceptAttribute(ConceptAttributeType conceptAttributeType) {
 		return dao.getConceptAttributeCount(conceptAttributeType) > 0;
 	}
+	
+	/**
+	 * @see org.openmrs.api.ConceptService#getConceptByReference(String conceptRef
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public Concept getConceptByReference(String conceptRef) {
+		if (StringUtil.isBlank(conceptRef)) {
+			return null;
+		}
+		Concept cpt = null;
+		//check if input is a valid Uuid
+		if (!(conceptRef.length() < 36 || conceptRef.length() > 38 || conceptRef.contains(" ") || conceptRef.contains("."))) {
+			cpt = Context.getConceptService().getConceptByUuid(conceptRef);
+			if (cpt != null) {
+				return cpt;
+			}
+		}
+		//handle mapping
+		int idx = conceptRef.indexOf(":");
+		if (idx >= 0 && idx < conceptRef.length() - 1) {
+			String conceptSource = conceptRef.substring(0, idx);
+			String conceptCode = conceptRef.substring(idx + 1);
+			conceptCode = conceptRef.substring(idx + 1);
+			conceptSource = conceptRef.substring(0, idx);
+			cpt = Context.getConceptService().getConceptByMapping(conceptCode, conceptSource);
+			if (cpt != null) {
+				return cpt;
+			}
+		}
+		//handle id
+		int conceptId = NumberUtils.toInt(conceptRef, -1);
+		if (conceptId >= 0) {
+			cpt = Context.getConceptService().getConcept(conceptId);
+			if (cpt != null) {
+				return cpt;
+			}
+		} else {
+			//handle name
+			cpt = Context.getConceptService().getConceptByName(conceptRef);
+			if (cpt != null) {
+				return cpt;
+			}
+		}
+		if (conceptRef.contains(".")) {
+			return getConceptByReference(evaluateStaticConstant(conceptRef));
+		}
```

Comment:
```suggestion
		if (conceptRef.contains(".")) {
		    try {
			    return getConceptByReference(evaluateStaticConstant(conceptRef));
			} catch (IllegalArgumentException e) {
			    log.error("Error while evaluating '{}' as a constant", conceptRef, e);
			}
		}
```

Reply:
@ibacher!  Thsnk you so much for making me realize that sometimes the static constant might be faulty.   And hence the `try-catch` is very necessary.

Reply:
Thank you @ibacher!  But i think we have a similar functionality in method `evaluateStaticConstant(String)` though we catch `Exception`.  I think any Error that occurs while Loading and retrieving the concept should shows that the static constant passed to the method was not correct, e.g miss-spelled,   

      private static String evaluateStaticConstant(String fqn) {
		......
		try {
			Class<?> clazz = Context.loadClass(clazzName);
			Field constantField = clazz.getField(constantName);
			Object val = constantField.get(null);
			return val != null ? String.valueOf(val) : null;
		}
		catch (Exception ex) {
			throw new IllegalArgumentException("Error while evaluating '{}' as a constant " + fqn, ex);
		}
	}